curl/src/main/scala/org/http4s/curl/internal/CurlEasy.scala

```diff
@@ -0,0 +1,130 @@
+    val code = thunk
+    if (code != 0) {
+      val info = fromCString(curl_easy_strerror(code))
+      val details = fromCString(errBuffer)
+      throw new CurlError(code, info, details)
+    }
+  }
+
+  def setUrl(URL: Ptr[CChar]): Unit = throwOnError(curl_easy_setopt_url(curl, CURLOPT_URL, URL))
+
+  def setCustomRequest(request: Ptr[CChar]): Unit =
+    throwOnError(curl_easy_setopt_customrequest(curl, CURLOPT_CUSTOMREQUEST, request))
+
+  def setHttpHeader(
+      headers: Ptr[curl_slist]
+  ): Unit = throwOnError(curl_easy_setopt_httpheader(curl, CURLOPT_HTTPHEADER, headers))
+
+  def setHttpVersion(
+      version: CLong
+  ): Unit = throwOnError(curl_easy_setopt_http_version(curl, CURLOPT_HTTP_VERSION, version))
+
+  def setHeaderFunction(
+      header_callback: header_callback
+  ): Unit = throwOnError(
+    curl_easy_setopt_headerfunction(curl, CURLOPT_HEADERFUNCTION, header_callback)
+  )
+
+  def setHeaderData(
+      pointer: Ptr[Byte]
+  ): Unit = throwOnError(curl_easy_setopt_headerdata(curl, CURLOPT_HEADERDATA, pointer))
+
+  def setWriteFunction(
+      write_callback: write_callback
+  ): Unit = throwOnError(
+    curl_easy_setopt_writefunction(curl, CURLOPT_WRITEFUNCTION, write_callback)
+  )
+
+  def setWriteData(
+      pointer: Ptr[Byte]
+  ): Unit = throwOnError(curl_easy_setopt_writedata(curl, CURLOPT_WRITEDATA, pointer))
+
+  def setReadFunction(
+      read_callback: read_callback
+  ): Unit = throwOnError(curl_easy_setopt_readfunction(curl, CURLOPT_READFUNCTION, read_callback))
+
+  def setReadData(
+      pointer: Ptr[Byte]
+  ): Unit = throwOnError(curl_easy_setopt_readdata(curl, CURLOPT_READDATA, pointer))
+
+  def setUpload(value: Boolean): Unit =
+    throwOnError(curl_easy_setopt_upload(curl, CURLOPT_UPLOAD, if (value) 1 else 0))
+
+  def setVerbose(value: Boolean): Unit =
+    throwOnError(curl_easy_setopt_verbose(curl, CURLOPT_VERBOSE, if (value) 1 else 0))
+
+  def setWebsocket(
+      flags: CLong
+  ): Unit = throwOnError(curl_easy_setopt_websocket(curl, CURLOPT_WS_OPTIONS, flags))
+
+  def wsSend(
+      buffer: Ptr[Byte],
+      bufLen: CSize,
+      send: Ptr[CSize],
+      fragsize: CSize,
+      flags: UInt,
+  ): Unit = throwOnError(curl_easy_ws_send(curl, buffer, bufLen, send, fragsize, flags))
+
+  def wsMeta(): Ptr[curl_ws_frame] = curl_easy_ws_meta(curl)
+
+  def pause(bitmask: CInt): Unit = throwOnError(curl_easy_pause(curl, bitmask))
+}
+
+private[curl] object CurlEasy {
+  private val CURL_ERROR_SIZE = 256L
+
+  private val createHandler: Resource[IO, Ptr[CURL]] = Resource.make {
+    IO {
+      val handle = curl_easy_init()
+      if (handle == null)
+        throw new RuntimeException("curl_easy_init")
+      handle
+    }
+  } { handle =>
+    IO(curl_easy_cleanup(handle))
+  }
+
+  def apply(): Resource[IO, CurlEasy] =
+    for {
+      h <- createHandler
+      z <- Utils.newZone
+      buf = z.alloc(CURL_ERROR_SIZE.toULong)
```

Comment:
Since this is a known, fixed size, instead of pre-allocating it, we can stack-allocate it whenever we need it in the `throwOnError` method.

Reply:
We need to pre-allocate this buffer, as it is used while setting options for the curl easy handler, and curl might populate it with human readable error if any, so it has a lifetime equal to curl handler itself.

Reply:
Ah, sorry, I missed that, thanks!